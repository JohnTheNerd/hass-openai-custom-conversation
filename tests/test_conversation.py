"""Tests for the vicuna_conversation component."""

from unittest.mock import AsyncMock, patch

import pytest
from syrupy.assertion import SnapshotAssertion
from openai.types.chat.chat_completion import ChatCompletion, Choice
from openai.types.chat.chat_completion_message import ChatCompletionMessage
from openai.types.completion_usage import CompletionUsage

from homeassistant.core import Context, HomeAssistant
from homeassistant.helpers import intent
from homeassistant.components import conversation

from pytest_homeassistant_custom_component.common import (
    MockConfigEntry,
)


@pytest.fixture(autouse=True)
def mock_setup_integration(config_entry: MockConfigEntry) -> None:
    """Setup the integration"""


async def test_conversation_entity(
    hass: HomeAssistant,
    config_entry: MockConfigEntry,
    snapshot: SnapshotAssertion,
) -> None:
    """Verify the conversation entity is loaded."""
    with patch(
        "openai.resources.chat.completions.AsyncCompletions.create",
        new_callable=AsyncMock,
        return_value=ChatCompletion(
            id="chatcmpl-1234567890ABCDEFGHIJKLMNOPQRS",
            choices=[
                Choice(
                    finish_reason="stop",
                    index=0,
                    message=ChatCompletionMessage(
                        content="Hello, how can I help you?",
                        role="assistant",
                        function_call=None,
                        tool_calls=None,
                    ),
                )
            ],
            created=1700000000,
            model="gpt-3.5-turbo-0613",
            object="chat.completion",
            system_fingerprint=None,
            usage=CompletionUsage(
                completion_tokens=9, prompt_tokens=8, total_tokens=17
            ),
        ),
    ) as mock_create:
        result = await conversation.async_converse(
            hass,
            "hello",
            None,
            Context(),
            agent_id=config_entry.entry_id,
        )

    assert result.response.response_type == intent.IntentResponseType.ACTION_DONE
    assert mock_create.mock_calls[0][2]["messages"] == snapshot
