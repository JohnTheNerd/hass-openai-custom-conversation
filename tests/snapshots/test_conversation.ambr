# serializer version: 1
# name: test_conversation_entity
  list([
    UserContent(role='user', content='hello'),
    AssistantContent(role='assistant', agent_id='conversation.mock_title', content='Hello, how can I help you?', tool_calls=None),
  ])
# ---
# name: test_function_call
  list([
    UserContent(role='user', content='Please call the test function'),
    AssistantContent(role='assistant', agent_id='conversation.mock_title', content=None, tool_calls=[ToolInput(tool_name='test_tool', tool_args={'param1': 'call1'}, id='call_call_1')]),
    ToolResultContent(role='tool_result', agent_id='conversation.mock_title', tool_call_id='call_call_1', tool_name='test_tool', tool_result='value1'),
    AssistantContent(role='assistant', agent_id='conversation.mock_title', content='I have successfully called the function', tool_calls=None),
  ])
# ---
# name: test_function_exception[]
  'Unexpected tool argument response: Expecting value: line 1 column 1 (char 0)'
# ---
# name: test_function_exception[{"para]
  'Unexpected tool argument response: Unterminated string starting at: line 1 column 2 (char 1)'
# ---
# name: test_unknown_hass_api
  dict({
    'continue_conversation': False,
    'conversation_id': 'conversation-id',
    'response': dict({
      'card': dict({
      }),
      'data': dict({
        'code': 'unknown',
      }),
      'language': 'en',
      'response_type': 'error',
      'speech': dict({
        'plain': dict({
          'extra_data': None,
          'speech': 'Error preparing LLM API',
        }),
      }),
    }),
  })
# ---
